cpp/CMakeLists.txt

```diff
@@ -51,9 +52,28 @@ set(root_directory ${PROJECT_BINARY_DIR})
 #
 string(TOLOWER ${CMAKE_BUILD_TYPE} LOWERCASE_BUILD_TYPE)
 if (${LOWERCASE_BUILD_TYPE} STREQUAL "debug" OR ${LOWERCASE_BUILD_TYPE} STREQUAL "relwithdebinfo")
+  set(CMAKE_BUILD_TYPE Debug)
+
   add_definitions(-DGLUTEN_PRINT_DEBUG)
+  message(STATUS "Add definition GLUTEN_PRINT_DEBUG")
+
+  set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -ggdb -O0")
+  message(STATUS "CMAKE_CXX_FLAGS_DEBUG=${CMAKE_CXX_FLAGS_DEBUG}")
+else ()
+  add_definitions(-DNDEBUG)
+  message(STATUS "Add definition NDEBUG")
 endif()
```

Comment:
Does the code bring new feature making `${CMAKE_BUILD_TYPE}` case-insensitive, or fix current issue from compile scripts?

Reply:
> Does the code bring new feature making `${CMAKE_BUILD_TYPE}` case-insensitive, or fix current issue from compile scripts?

The `cmake` tutorial tells us to use `Debug` `Release` rather than `debug release`. refer to [CMAKE_BUILD_TYPE](https://cmake.org/cmake/help/v3.20/variable/CMAKE_BUILD_TYPE.html)

In addition, I set `-ggdb -O0` for `CMAKE_CXX_FLAGS_DEBUG` to make debugging better.

Reply:
Thanks for the explanation.

But why not to just check if `${CMAKE_BUILD_TYPE} STREQUAL "Debug"`, any special reasons?

Reply:
Does that mean the value of `${CMAKE_BUILD_TYPE}` is treat as case-insensitive by CMake? e.g. `-DCMAKE_BUILD_TYPE=ReLeAsE` is a legal writting?

Reply:
> Thanks for the explanation.
> 
> But why not to just check if `${CMAKE_BUILD_TYPE} STREQUAL "Debug"` (and `RelWithDebInfo`), any special reasons?

this PR does not make this change of `RelWithDebInfo`

Reply:
> Does that mean the value of `${CMAKE_BUILD_TYPE}` is treated as case-insensitive by CMake? e.g. `-DCMAKE_BUILD_TYPE=ReLeAsE` is a legal writing?

I am not sure, I can not find related content about this, using `Debug Release` is legal according to the tutorial.

Reply:
![image](https://user-images.githubusercontent.com/14065538/209746231-4d767892-22aa-4536-867c-77d5e822a972.png)
-DCMAKE_BUILD_TYPE=ReLeAsE is a legal writing

Reply:
That's OK to me if the code was historical, we may still need to check it's correctness in future.

> -DCMAKE_BUILD_TYPE=ReLeAsE is a legal writing

This doesn't seem to be complete evidence since `-DCMAKE_BUILD_TYPE=Release` can sometime be set as default.

By the way, do we have to rewrite `RelWithDebInfo` to `Debug` here? Sometimes `RelWithDebInfo` can be useful too.

Reply:
something like:

```suggestion
string(TOLOWER ${CMAKE_BUILD_TYPE} LOWERCASE_BUILD_TYPE)
if (${LOWERCASE_BUILD_TYPE} STREQUAL "relwithdebinfo")
  add_definitions(-DGLUTEN_PRINT_DEBUG) // this is the original logic, right?
  message(STATUS "Add definition GLUTEN_PRINT_DEBUG")
endif()

if (${LOWERCASE_BUILD_TYPE} STREQUAL "debug" OR ${LOWERCASE_BUILD_TYPE} STREQUAL "relwithdebinfo")
  add_definitions(-DGLUTEN_PRINT_DEBUG)
  message(STATUS "Add definition GLUTEN_PRINT_DEBUG")
  set(CMAKE_BUILD_TYPE Debug)
  set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -ggdb -O0")
  message(STATUS "CMAKE_CXX_FLAGS_DEBUG=${CMAKE_CXX_FLAGS_DEBUG}")
else ()
  add_definitions(-DNDEBUG)
  message(STATUS "Add definition NDEBUG")
endif()
```

Reply:
OK，I will modify it as above